gax/src/main/java/com/google/api/gax/batching/BatcherImpl.java

```diff
@@ -102,15 +107,54 @@ public BatcherImpl(
       BatchingSettings batchingSettings,
       ScheduledExecutorService executor) {
 
+    this(batchingDescriptor, unaryCallable, prototype, batchingSettings, executor, null);
+  }
+
+  /**
+   * @param batchingDescriptor a {@link BatchingDescriptor} for transforming individual elements
+   *     into wrappers request and response
+   * @param unaryCallable a {@link UnaryCallable} object
+   * @param prototype a {@link RequestT} object
+   * @param batchingSettings a {@link BatchingSettings} with configuration of thresholds
+   * @param flowController a {@link FlowController} for throttling requests. If it's null, create a
+   *     {@link FlowController} object from {@link BatchingSettings#getFlowControlSettings()}.
+   */
+  public BatcherImpl(
+      BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT> batchingDescriptor,
+      UnaryCallable<RequestT, ResponseT> unaryCallable,
+      RequestT prototype,
+      BatchingSettings batchingSettings,
+      ScheduledExecutorService executor,
+      @Nullable FlowController flowController) {
+
     this.batchingDescriptor =
         Preconditions.checkNotNull(batchingDescriptor, "batching descriptor cannot be null");
     this.unaryCallable = Preconditions.checkNotNull(unaryCallable, "callable cannot be null");
     this.prototype = Preconditions.checkNotNull(prototype, "request prototype cannot be null");
     this.batchingSettings =
         Preconditions.checkNotNull(batchingSettings, "batching setting cannot be null");
     Preconditions.checkNotNull(executor, "executor cannot be null");
+    if (flowController == null) {
+      flowController = new FlowController(batchingSettings.getFlowControlSettings());
```

Comment:
This is suspicious, as it lets contradicting parameters to be passed in the constructor. Specifically `batchingSettings` has `flowControlSettings` in it, which should be respected. Here we basically let overwrite it with directly specifying `flowController`. It makes it hard to keep things under control: what is even the meaning of `batchingSettigns.getFlowControlSettings()` if `flowController` was specified explicitly? 

Why do we need pass `FlowController` directly? Was it added for testing only? If it was added for testing only. We should avoid damaging surface for test purposes only (in worst case create a package-private constructor (which will accept `FlowController` directly) with VisibleForTesting annotation and proper comment.

Reply:
This is similar to the implementation of `BathcingCallSettings` constructor (https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/rpc/BatchingCallSettings.java#L70-L72) and `BatcherFactory` constructor (https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/rpc/BatcherFactory.java).

In my opinion, `FlowControlSettings` maybe shouldn't be part of `BatchingSettings`. If we need to create multiple `Batcher` instances, we'll want them to share the same `FlowController`, so we do need a way to pass FlowController directly.

Reply:
Oh... so it seems like it is already messed up =(. Ok